### PR TITLE
Integrating LIEF parser for SHARED builds

### PR DESCRIPTION
**Pull request by prikhap**
Title: Add CI workflow for C++ inject_hash implementation

Description:
This PR adds CI workflow and testing infrastructure for the C++ inject_hash implementation.

Changes:
- Add GitHub Actions workflow for testing
- Add test script for building and verifying functionality
- Test on both Ubuntu and macOS environments

Testing:
- Verified workflow locally
- Tests build process
- Tests FIPS integrity verification

### CODE REVIEW

File: util/fipstools/inject_hash_cpp/test_sha256.cpp
Comment (samuel40791765) [inline util/fipstools/inject_hash_cpp/test_sha256.cpp:1]: Is this file used right now? I understand that it might come up in later implementations, but we can leave until then to pull this in.

File: tests/ci/run_inject_hash_cpp.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+fi
+
+# First try to build just inject_hash_cpp
+echo "Building inject_hash_cpp..."
+ninja inject_hash_cpp
+
+# Test inject_hash_cpp basic functionality
+echo "Testing inject_hash_cpp basic functionality..."
+./util/fipstools/inject_hash_cpp/inject_hash_cpp --version || echo "Version check failed"
+./util/fipstools/inject_hash_cpp/inject_hash_cpp --help || echo "Help check failed"
```
Comment (samuel40791765) [inline tests/ci/run_inject_hash_cpp.sh]: This doesn't seem to be used, could we remove this?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,31 @@
 cmake_minimum_required(VERSION 3.5..3.31)
 
+
+####################
```
Comment (samuel40791765) [inline CMakeLists.txt]: The comment/text here is redundant.

File: tests/ci/run_inject_hash_cpp.sh [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+# Build everything
+echo "Building full project..."
+ninja
+
+# Test actual FIPS injection
```
Comment (samuel40791765) [inline tests/ci/run_inject_hash_cpp.sh]: Redundant comment here as well.

File: tests/ci/run_inject_hash_cpp.sh [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        -DFIPS_SHARED=1 \
+        -DBUILD_SHARED_LIBS=1 \
+        -DUSE_CPP_INJECT_HASH=ON \
+        -DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk \
+        ..
+else
+    cmake -GNinja \
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo \
+        -DFIPS=1 \
+        -DFIPS_SHARED=1 \
```
Comment (samuel40791765) [inline tests/ci/run_inject_hash_cpp.sh]: We should also be testing the shared build with ubuntu. And we should use `-DCMAKE_BUILD_TYPE=Release` since that's what we validate for FIPS :)

File: tests/ci/run_inject_hash_cpp.sh [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        -DBUILD_SHARED_LIBS=1 \
+        -DUSE_CPP_INJECT_HASH=ON \
+        ..
+fi
+
+# First try to build just inject_hash_cpp
+echo "Building inject_hash_cpp..."
+ninja inject_hash_cpp
```
Comment (samuel40791765) [inline tests/ci/run_inject_hash_cpp.sh]: I don't think `inject_hash_cpp` is necessarily used here. Calling `ninja` with the build flags you've set up should be sufficient enough.

We could also probably drop the `ninja` call on L49 below in favor of this line.

File: util/fipstools/inject_hash_cpp/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        ${CMAKE_CURRENT_BINARY_DIR}
+        ${CMAKE_SOURCE_DIR}/include
+    )
+
+    if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
+        target_compile_options(inject_hash_cpp PRIVATE
+            -Wno-overloaded-virtual
+            -Wno-unused-parameter
+        )
+    endif()
```
Comment (samuel40791765) [inline util/fipstools/inject_hash_cpp/CMakeLists.txt:21]: Are these additional linker flags needed? AWS-LC already enables all warning linker flags in the build, so this could be redundant.

File: util/fipstools/inject_hash_cpp/inject_hash.cpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        }
+    } else {
+        std::cerr << "❌ No binary path provided" << std::endl;
+        return 1;
+    }
+
+    std::cout << "\nReceived arguments:" << std::endl;
+    for (int i = 1; i < argc; i++) {
+        std::cout << "  Arg " << i << ": " << argv[i] << std::endl;
+    }
```
Comment (samuel40791765) [inline util/fipstools/inject_hash_cpp/inject_hash.cpp]: This is probably useful for debugging, but we should drop it since we don't want debugging prints in most production code.

File: tests/ci/run_inject_hash_cpp.sh [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+./util/fipstools/inject_hash_cpp/inject_hash_cpp --help || echo "Help check failed"
+
+# Build everything
+echo "Building full project..."
+ninja
+
+# Test actual FIPS injection
+echo "Testing FIPS injection..."
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    ./util/fipstools/inject_hash_cpp/inject_hash_cpp -o ./crypto/libcrypto.dylib -in-object ./crypto/libcrypto.dylib -apple
```
Comment (samuel40791765) [inline tests/ci/run_inject_hash_cpp.sh]: `-o ./crypto/libcrypto.dylib` and `-apple` don't seem to actually be used here. Same for below.

File: util/fipstools/inject_hash_cpp/inject_hash.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    std::cout << "\n=== C++ inject_hash starting ===" << std::endl;
+    std::cout << "Testing LIEF integration..." << std::endl;
+
+    const char* binary_path = nullptr;
+    for (int i = 1; i < argc - 1; i++) {
+        if (strcmp(argv[i], "-in-object") == 0) {
+            binary_path = argv[i + 1];
+            break;
+        }
+    }
```
Comment (samuel40791765) [inline util/fipstools/inject_hash_cpp/inject_hash.cpp]: There's probably a more graceful way to handle this. Could you look into pulling in the argument parsing functions in `tool/internal.h`?
* https://github.com/aws/aws-lc/blob/main/tool/internal.h#L107-L114

`ParseKeyValueArguments` should help you parse what you need in a cleaner and more easily extendable fashion.

File: tests/ci/run_inject_hash_cpp.sh (hunk 4 above)
Comment (prikhap) [reply tests/ci/run_inject_hash_cpp.sh]: static *

File: util/fipstools/inject_hash_cpp/inject_hash.cpp (hunk 9 above)
Comment (prikhap) [reply util/fipstools/inject_hash_cpp/inject_hash.cpp]: I just kept this for testing purpose. will keep this in mind while making the final file

File: util/fipstools/inject_hash_cpp/CMakeLists.txt (hunk 6 above)
Comment (prikhap) [reply util/fipstools/inject_hash_cpp/CMakeLists.txt:21]: LIEF has it's own warning, which were resulting in an error. Therefore used these flags to disable -WError

File: tests/ci/run_inject_hash_cpp.sh [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+echo "Environment:"
+go version
+cmake --version
+ninja --version
+
+# Build AWS-LC with C++ inject_hash
+mkdir -p build
+cd build
+
+# Configure based on build type and OS
```
Comment (andrewhop) [inline tests/ci/run_inject_hash_cpp.sh]: Nothing is actually building aws-lc, this is only running CMake to configure the build.

File: tests/ci/run_inject_hash_cpp.sh [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    # macOS only supports shared builds
+    cmake -GNinja \
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo \
+        -DFIPS=1 \
+        -DFIPS_SHARED=1 \
+        -DBUILD_SHARED_LIBS=1 \
+        -DUSE_CPP_INJECT_HASH=ON \
+        -DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk \
+        ..
```
Comment (andrewhop) [inline tests/ci/run_inject_hash_cpp.sh]: There are existing helper functions defined in `common_posix_setup.sh` to help simplify this.

File: util/fipstools/inject_hash_cpp/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+    target_link_libraries(inject_hash_cpp PRIVATE
+        LIEF::LIEF
+    )
+
+    set_target_properties(inject_hash_cpp PROPERTIES
+        CXX_STANDARD 17
+        CXX_STANDARD_REQUIRED ON
+    )
```
Comment (andrewhop) [inline util/fipstools/inject_hash_cpp/CMakeLists.txt]: Why 17?

File: util/fipstools/inject_hash_cpp/inject_hash.cpp [hunk 13]
```diff
@@ -0,0 +1,34 @@
+#include <LIEF/LIEF.hpp>
+#include <iostream>
+#include <string>
+
+int main(int argc, char** argv) {
+    std::cout << "\n=== C++ inject_hash starting ===" << std::endl;
```
Comment (andrewhop) [inline util/fipstools/inject_hash_cpp/inject_hash.cpp:20]: Nothing appears to be running this, I don't see it in any of the GitHub action output.

File: .github/workflows/test-cpp-inject-hash.yml [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+      if: runner.os == 'macOS'
+      run: |
+        brew install ninja cmake perl go
+        
+    - name: Build and Test
+      env:
+        BUILD_TYPE: ${{ matrix.build-type }}
+      run: |
+        chmod +x tests/ci/run_inject_hash_cpp.sh
+        ./tests/ci/run_inject_hash_cpp.sh
```
Comment (andrewhop) [inline .github/workflows/test-cpp-inject-hash.yml]: Why not check the bash script in with the correct file permissions?

File: util/fipstools/inject_hash_cpp/inject_hash.cpp (hunk 13 above)
Comment (samuel40791765) [reply util/fipstools/inject_hash_cpp/inject_hash.cpp:20]: I think it's not running because the last few lines in `tests/ci/run_inject_hash_cpp.sh` were accidentally commented out.

File: util/fipstools/inject_hash_cpp/test_sha256.cpp
Comment (prikhap) [reply util/fipstools/inject_hash_cpp/test_sha256.cpp:1]: Removed it!

File: tests/ci/run_inject_hash_cpp.sh (hunk 1 above)
Comment (prikhap) [reply tests/ci/run_inject_hash_cpp.sh]: Removed it!

File: tests/ci/run_inject_hash_cpp.sh (hunk 10 above)
Comment (prikhap) [reply tests/ci/run_inject_hash_cpp.sh]: missed adding the ninja command.

File: util/fipstools/inject_hash_cpp/CMakeLists.txt (hunk 12 above)
Comment (prikhap) [reply util/fipstools/inject_hash_cpp/CMakeLists.txt]: Had kept it since LIEF requires this version of C++. However, we don't have to bother about the version while building aws-lc, hence deleted the properties from CMake.

File: .github/workflows/test-cpp-inject-hash.yml (hunk 14 above)
Comment (prikhap) [reply .github/workflows/test-cpp-inject-hash.yml]: done

File: util/fipstools/inject_hash_cpp/inject_hash.cpp (hunk 13 above)
Comment (prikhap) [reply util/fipstools/inject_hash_cpp/inject_hash.cpp:20]: I mistakenly commented commands to run inject_hash, should work now.

File: tests/ci/run_inject_hash_cpp.sh (hunk 4 above)
Comment (prikhap) [reply tests/ci/run_inject_hash_cpp.sh]: Tried handling static files, which require creating an archive file parser (left it for some other PR)

File: tests/ci/run_inject_hash_cpp.sh (hunk 5 above)
Comment (prikhap) [reply tests/ci/run_inject_hash_cpp.sh]: Done

File: util/fipstools/inject_hash_cpp/inject_hash.cpp (hunk 7 above)
Comment (prikhap) [reply util/fipstools/inject_hash_cpp/inject_hash.cpp]: the inject_hash.cpp currently is just a test file, added some test functionalities and std:cout to see if the file was actually called.

File: tests/ci/run_inject_hash_cpp.sh (hunk 8 above)
Comment (prikhap) [reply tests/ci/run_inject_hash_cpp.sh]: Modified according to the CMake compiler flags

File: tests/ci/run_inject_hash_cpp.sh (hunk 11 above)
Comment (prikhap) [reply tests/ci/run_inject_hash_cpp.sh]: included in the recent commit.

File: CMakeLists.txt [hunk 15]
```diff
@@ -1,5 +1,28 @@
...
+    set(LIEF_DOC OFF CACHE BOOL "")
+    set(LIEF_INSTALL ON CACHE BOOL "")
+    set(LIEF_STATIC ON CACHE BOOL "")
+    set(LIEF_ELF ON CACHE BOOL "Enable ELF format")
+    set(LIEF_PE ON CACHE BOOL "Enable PE format")
+    set(LIEF_MACHO ON CACHE BOOL "Enable MACHO format")
+    set(LIEF_DEX OFF CACHE BOOL "Disable DEX format")
+    set(LIEF_VDEX OFF CACHE BOOL "Disable VDEX format")
+    set(LIEF_ART OFF CACHE BOOL "Disable ART format")
+    set(LIEF_OAT OFF CACHE BOOL "Disable OAT format")
```
Comment (andrewhop) [inline CMakeLists.txt]: It is much more useful if your comment explains what these options are and why you chose these specific values.

File: CMakeLists.txt [hunk 16]
```diff
@@ -1,5 +1,28 @@
...
+    
+    # Add LIEF subdirectory
+    add_subdirectory(third_party/lief)
```
Comment (andrewhop) [inline CMakeLists.txt]: It is not necessary to comment every line, especially when they are self explanatory like this. 
```suggestion
    add_subdirectory(third_party/lief)
```

File: CMakeLists.txt [hunk 17]
```diff
@@ -1,5 +1,28 @@
...
+    
+    # Make LIEF headers available
+    include_directories(${CMAKE_CURRENT_SOURCE_DIR}/third_party/lief/include)
```
Comment (andrewhop) [inline CMakeLists.txt]: This makes the LIEF headers available to every target, can you scope this down to just the inject hash tool?

File: crypto/CMakeLists.txt [hunk 18]
```diff
@@ -652,14 +652,28 @@ if(FIPS_SHARED)
       set(INJECT_HASH_APPLE_FLAG "-apple")
     endif()
 
+    if(USE_CPP_INJECT_HASH)
+    # Add dependency on C++ implementation
```
Comment (andrewhop) [inline crypto/CMakeLists.txt]: Indentation looks off for the if and else block, more superfluous comments

File: crypto/CMakeLists.txt [hunk 19]
```diff
@@ -652,14 +652,28 @@ if(FIPS_SHARED)
...
+    add_dependencies(crypto inject_hash_cpp)
+    
+    # C++ implementation for hash injection
+    add_custom_command(
+      TARGET crypto POST_BUILD
+      COMMAND ${CMAKE_BINARY_DIR}/util/fipstools/inject_hash_cpp/inject_hash_cpp
+      -o $<TARGET_FILE:crypto>
+      -in-object $<TARGET_FILE:crypto>
+      ${INJECT_HASH_APPLE_FLAG}
```
Comment (andrewhop) [inline crypto/CMakeLists.txt]: What are you using this flag for?

File: crypto/fipsmodule/CMakeLists.txt [hunk 20]
```diff
@@ -389,15 +389,26 @@ if(FIPS_DELOCATE)
...
+#  if(USE_CPP_INJECT_HASH)
+#     add_custom_command(
+#         OUTPUT bcm.o
+#         COMMAND ../../util/fipstools/inject_hash_cpp/inject_hash_cpp
+#         -o bcm.o
+#         -in-archive $<TARGET_FILE:bcm_hashunset>
+#         DEPENDS bcm_hashunset inject_hash_cpp
+#         WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+#     )
+# else()
```
Comment (andrewhop) [inline crypto/fipsmodule/CMakeLists.txt]: I prefer not checking in unused/commented out code. It's likely to get out of date and require rewriting in the future.
```suggestion
# else()
```

File: tests/ci/run_inject_hash_cpp.sh [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+# Source common setup
+source tests/ci/common_posix_setup.sh
+
+# Build AWS-LC with C++ inject_hash
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    # macOS only supports shared builds
+    run_build \
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo \
+        -DFIPS=1 \
+        -DFIPS_SHARED=1 \
```
Comment (andrewhop) [inline tests/ci/run_inject_hash_cpp.sh]: Our CMakeList figures this out automatically
```suggestion
```

File: tests/ci/run_inject_hash_cpp.sh [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+        -DBUILD_SHARED_LIBS=1 \
+        -DUSE_CPP_INJECT_HASH=ON \
+        -DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
```
Comment (andrewhop) [inline tests/ci/run_inject_hash_cpp.sh]: Why are you specifying this?

File: crypto/fipsmodule/CMakeLists.txt [hunk 23]
```diff
@@ -389,15 +389,26 @@ if(FIPS_DELOCATE)
...
-
-  go_executable(inject_hash
-                boringssl.googlesource.com/boringssl/util/fipstools/inject_hash)
-  add_custom_command(
-    OUTPUT bcm.o
-    COMMAND ./inject_hash -o bcm.o -in-archive $<TARGET_FILE:bcm_hashunset>
-    DEPENDS bcm_hashunset inject_hash
-    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
-  )
+  #ignoring static builds for now
```
Comment (samuel40791765) [inline crypto/fipsmodule/CMakeLists.txt]: It would be helpful to document the reason in the comments, but we can do that in the next PR.

**Review by samuel40791765 — APPROVED** [review]

File: tests/ci/run_inject_hash_cpp.sh [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+        -DBUILD_SHARED_LIBS=1 \
+        -DUSE_CPP_INJECT_HASH=ON
+    # fi
+fi
+
+cd "${BUILD_ROOT}"
+
+# Test actual FIPS injection
+echo "TESTING INJECT_HASH.CPP WITH EDGE CASES..." 
+./util/fipstools/inject_hash_cpp/inject_hash_cpp || echo "--Expected failure with no args--"
```
Comment (andrewhop) [inline tests/ci/run_inject_hash_cpp.sh]: Should this fail the build if this isn't an error? Also I prefer moving more of this test logic into actual test files like our tools.

File: tests/ci/run_inject_hash_cpp.sh [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    ./util/fipstools/inject_hash_cpp/inject_hash_cpp \
+        -o ./crypto/libcrypto.dylib \
+        -in-object ./crypto/libcrypto.dylib \
+        -apple
+else
+    ./util/fipstools/inject_hash_cpp/inject_hash_cpp \
+        -o ./crypto/libcrypto.so \
+        -in-object ./crypto/libcrypto.so
+fi
```
Comment (andrewhop) [inline tests/ci/run_inject_hash_cpp.sh]: Isn't this run (with the correct paths) automatically during the normal FIPS build?

File: util/fipstools/inject_hash_cpp/inject_hash.cpp [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+            i++; // Skip next argument
+        }
+        else if (strcmp(argv[i], "-apple") == 0) {
+            is_apple = true;
+        }
+    }
+
+    // Validate arguments
+    if (!input_file || !output_file) {
+        std::cerr << "❌ Missing required arguments" << std::endl;
```
Comment (andrewhop) [inline util/fipstools/inject_hash_cpp/inject_hash.cpp]: It is more helpful to say which ones are missing to help future customers/developers debug the issue.

File: util/fipstools/inject_hash_cpp/inject_hash.cpp [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+            input_file = argv[i + 1];
+            i++; // Skip next argument
+        }
+        else if (strcmp(argv[i], "-o") == 0) {
+            output_file = argv[i + 1];
+            i++; // Skip next argument
+        }
+        else if (strcmp(argv[i], "-apple") == 0) {
+            is_apple = true;
+        }
```
Comment (andrewhop) [inline util/fipstools/inject_hash_cpp/inject_hash.cpp]: Why do we need to handle Apple in a special way?

File: util/fipstools/inject_hash_cpp/inject_hash.cpp [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+    std::cout << "Usage: inject_hash_cpp -o output_file -in-object input_file [-apple]" << std::endl;
+}
+
+int main(int argc, char** argv) {
+    std::cout << "\n=== C++ inject_hash starting ===" << std::endl;
+
+    // Parse command line arguments
+    const char* input_file = nullptr;
+    const char* output_file = nullptr;
+    bool is_apple = false;
```
Comment (andrewhop) [inline util/fipstools/inject_hash_cpp/inject_hash.cpp]: Is it possible to re-use any of our existing argument parsing from our tools?

File: CMakeLists.txt (hunk 15 above)
Comment (prikhap) [reply CMakeLists.txt]: check LIEF_BUILD_OPTIONS at : https://lief.re/doc/latest/installation.html#cmake-integration

File: CMakeLists.txt (hunk 17 above)
Comment (prikhap) [reply CMakeLists.txt]: LIEF headers are scoped to inject_hash_cpp using the util/fipstools/inject_hash_cpp/CMakeLists.txt. I added it in the global CMake, to make the LIEF available throughout, but it surely can be removed and it is better directed in it's desired scope.

File: crypto/CMakeLists.txt (hunk 18 above)
Comment (prikhap) [reply crypto/CMakeLists.txt]: amdended!

File: crypto/CMakeLists.txt (hunk 19 above)
Comment (prikhap) [reply crypto/CMakeLists.txt]: these flags currently aren't of much usage, but was trying maintain similarity with the flag format used by the inject_hash.go script. Currently these flags are just printed as a part of testing the integration of the C++ script, however while creating a proper parsing and hashing script, these flags would be required to have the input path of the binary, output path and if it is suited for apple or linux.

File: tests/ci/run_inject_hash_cpp.sh (hunk 25 above)
Comment (prikhap) [reply tests/ci/run_inject_hash_cpp.sh]: this was added to the test script just to show contrast between "expected failure" cases and a normal case.

File: util/fipstools/inject_hash_cpp/inject_hash.cpp (hunk 27 above)
Comment (prikhap) [reply util/fipstools/inject_hash_cpp/inject_hash.cpp]: It doesen't serve any purpose right now, was just setting foundations for the future builds.

File: util/fipstools/inject_hash_cpp/inject_hash.cpp (hunk 26 above)
Comment (prikhap) [reply util/fipstools/inject_hash_cpp/inject_hash.cpp]: modified.

File: crypto/fipsmodule/CMakeLists.txt (hunk 20 above)
Comment (prikhap) [reply crypto/fipsmodule/CMakeLists.txt]: modified.

File: util/fipstools/inject_hash_cpp/inject_hash.cpp (hunk 28 above)
Comment (samuel40791765) [reply util/fipstools/inject_hash_cpp/inject_hash.cpp]: I also brought this up in https://github.com/aws/aws-lc/pull/2534#discussion_r2202011937. The answer seemed to imply that this would part of a larger future refactor/PR?

File: CMakeLists.txt [hunk 29]
```diff
@@ -1,5 +1,32 @@
...
+if(USE_CPP_INJECT_HASH)
+    # Configure LIEF build options first :  check LIEF_BUILD_OPTIONS at : https://lief.re/doc/latest/installation.html#cmake-integration
+    set(LIEF_PYTHON_API OFF CACHE BOOL "") # disabling LIEF's Python API
+    # Disabling examples and tests as we only need the core library
+    set(LIEF_EXAMPLES OFF CACHE BOOL "") 
+    set(LIEF_TESTS OFF CACHE BOOL "")
+    # disabling LIEF's documentation generation
+    set(LIEF_DOC OFF CACHE BOOL "")
+    # enabling installation of LIEF
+    set(LIEF_INSTALL ON CACHE BOOL "")
```
Comment (andrewhop) [inline CMakeLists.txt]: Why do we need to install LIEF?

File: crypto/CMakeLists.txt [hunk 30]
```diff
@@ -652,14 +652,27 @@ if(FIPS_SHARED)
...
-      TARGET crypto POST_BUILD
-      COMMAND ${GO_EXECUTABLE} run
-      ${AWSLC_SOURCE_DIR}/util/fipstools/inject_hash/inject_hash.go
-      -o $<TARGET_FILE:crypto> -in-object $<TARGET_FILE:crypto> ${INJECT_HASH_APPLE_FLAG}
-      WORKING_DIRECTORY ${AWSLC_SOURCE_DIR}
-    )
-
+    if(USE_CPP_INJECT_HASH)
+      # Add dependency on C++ implementation
+      add_dependencies(crypto inject_hash_cpp)
```
Comment (andrewhop) [inline crypto/CMakeLists.txt]: Do you think this needs a comment?

File: crypto/fipsmodule/CMakeLists.txt [hunk 31]
```diff
@@ -389,15 +389,17 @@ if(FIPS_DELOCATE)
...
+    #ignoring static builds for now as archive files require a seperate parser to extract .o files from it. The lief parser does not support archive files.
+    go_executable(inject_hash
+        boringssl.googlesource.com/boringssl/util/fipstools/inject_hash)
+    add_custom_command(
+        OUTPUT bcm.o
+        COMMAND ./inject_hash -o bcm.o -in-archive $<TARGET_FILE:bcm_hashunset>
+        DEPENDS bcm_hashunset inject_hash
+        WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+    )
+# endif()
```
Comment (andrewhop) [inline crypto/fipsmodule/CMakeLists.txt]: Besides the comment did anything change here? If not we shouldn't change any of the whitespace because it makes tracking actual changes in git history harder.

File: tests/ci/run_inject_hash_cpp.sh [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+        -DFIPS=1 \
+        -DBUILD_SHARED_LIBS=1 \
+        -DUSE_CPP_INJECT_HASH=ON 
+else
+    run_build \
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo \
+        -DFIPS=1 \
+        -DBUILD_SHARED_LIBS=1 \
+        -DUSE_CPP_INJECT_HASH=ON
+fi
```
Comment (andrewhop) [inline tests/ci/run_inject_hash_cpp.sh]: Is there any difference between the if and else block?

**Review by andrewhop — APPROVED** [review]
